packages/framework/src/Component/Translation/ConstraintViolationExtractor.php

```diff
@@ -0,0 +1,190 @@
+
+use JMS\TranslationBundle\Model\FileSource;
+use JMS\TranslationBundle\Model\Message;
+use JMS\TranslationBundle\Model\MessageCatalogue;
+use JMS\TranslationBundle\Translation\Extractor\FileVisitorInterface;
+use PhpParser\Node;
+use PhpParser\Node\Expr\MethodCall;
+use PhpParser\Node\Expr\Variable;
+use PhpParser\Node\Name\FullyQualified;
+use PhpParser\Node\Scalar\String_;
+use PhpParser\Node\Stmt\ClassMethod;
+use PhpParser\NodeTraverser;
+use PhpParser\NodeVisitor;
+use PhpParser\NodeVisitor\NameResolver;
+use SplFileInfo;
+use Symfony\Component\Validator\Context\ExecutionContextInterface;
+use Twig_Node;
+
+class ConstraintViolationExtractor implements FileVisitorInterface, NodeVisitor
+{
+    private const CONSTRAINT_MESSAGE_METHOD_NAME = 'addViolation';
+
+    /**
+     * @var \PhpParser\NodeTraverser
+     */
+    private $traverser;
+
+    /**
+     * @var \JMS\TranslationBundle\Model\MessageCatalogue
+     */
+    private $catalogue;
+
+    /**
+     * @var \SplFileInfo
+     */
+    private $file;
+
+    /**
+     * @var string
+     */
+    private $interfaceInstanceName;
+
+    public function __construct()
+    {
+        $this->traverser = new NodeTraverser();
+        $this->traverser->addVisitor(new NameResolver());
+        $this->traverser->addVisitor($this);
+        $this->interfaceInstanceName = '';
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function visitPhpFile(SplFileInfo $file, MessageCatalogue $catalogue, array $ast)
+    {
+        $this->file = $file;
+        $this->catalogue = $catalogue;
+        $this->traverser->traverse($ast);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function enterNode(Node $node)
+    {
+        if ($node instanceof ClassMethod && $this->haveMethodContextInterfaceImplemented($node) === true) {
+            foreach ($node->stmts as $stmt) {
+                $this->recursiveFindMessages($stmt);
+            }
+        }
+    }
+
+    /**
+     * @param \PhpParser\Node\Stmt\ClassMethod $node
+     * @return bool
+     */
+    private function haveMethodContextInterfaceImplemented(ClassMethod $node)
+    {
+        foreach ($node->getParams() as $param) {
+            if ($this->haveParamExecutionContextInterface($param) === true) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * @param \PhpParser\Node\Param $param
+     * @return bool
+     */
+    private function haveParamExecutionContextInterface(Node\Param $param)
+    {
+        if ($param->type instanceof FullyQualified) {
+            $className = implode('\\', $param->type->parts);
+            $interfaceName = ExecutionContextInterface::class;
+            $this->interfaceInstanceName = $param->name;
+
+            return $className === $interfaceName;
+        }
+
+        return false;
+    }
+
+    /**
+     * @param \PhpParser\Node $node
+     */
+    private function recursiveFindMessages(Node $node)
```

Comment:
It is not necessary to use the recursion at all as there is a component (`$traverser`) that helps you to achieve the same goal...take a look at the suggested refactoring